File: cliente.py
import Pyro5.api
import logging
import threading
import sys
import os
from Crypto.PublicKey import RSA
from Crypto.Signature import pkcs1_15
from Crypto.Hash import SHA256
from Crypto.Cipher import PKCS1_v1_5 as Cipher_PKCS1_v1_5

"""Classe cliente, que é responsável por fazer a comunicação com o servidor de nomes e com o servidor de estoque
    Atributos:
        nome: nome do cliente
        uriCliente: uri do cliente
    Métodos:
        pedeCriar: pede para criar o cliente
        inicializaDaemon: inicializa o daemon do cliente
        notificacao: notifica o cliente de algum acontecimento
            notificação é exposta para o servidor, para que possa acontecer a comunicação"""

class cliente():
    nome = ""
    uriCliente = ""

    def __init__(self):
        self.nome = input(
            "Bem vindo ao Gerenciador de Estoque!\n\nInforme seu nome de registro no servidor: ")

    def pedeCriar(self, uriCliente):
        self.uriCliente = uriCliente

    def inicializaDaemon(self, daemon):
        daemon.requestLoop()

    @Pyro5.api.expose
    @Pyro5.api.callback
    def notificacao(self, acontecimento):
        print(acontecimento)


if __name__ == '__main__':
    
    clienteInstancia = cliente()
    daemon = Pyro5.api.Daemon()
    uriCliente = daemon.register(clienteInstancia)
    clienteInstancia.pedeCriar(uriCliente)

    # Inicializa o daemon em uma thread separada para que o cliente possa receber notificações e não fique travado esperando resposta do servidor
    daemonThread = threading.Thread(
        target=clienteInstancia.inicializaDaemon, args=(daemon,), daemon=True)
    daemonThread.start()

    servidorNomes = Pyro5.api.locate_ns()
    uriGerenciadorEstoque = servidorNomes.lookup("Gerenciador de Estoque")
    servidorGerenciadorEstoque = Pyro5.api.Proxy(uriGerenciadorEstoque)

    print(servidorGerenciadorEstoque.registrarCliente(
        clienteInstancia.nome, clienteInstancia.uriCliente))
    
    # Interface do cliente
    while (1):
        print("\nAs opções do servidor são:")
        print("1 - Cadastrar novo produto")
        print("2 - Adicionar produto existente")
        print("3 - Retirar produto")
        print("4 - Listar produtos")
        print("5 - Gerar relatórios\n")
        opcao = input()

        if opcao == '1':
            codigoProduto = input("\nQual o código do produto? ")
            nomeProduto = input("Qual nome do produto? ")
            descricaoProduto = input("Qual a descrição do produto? ")
            quantidadeProduto = input("Qual a quantidade do produto? ")
            precoUnidadeProduto = input("Qual o preço por unidade do produto? ")
            estoqueMinimoProduto = input("Qual o estoque mínimo do produto? ")

            retorno = servidorGerenciadorEstoque.cadastrarProdutoNovo(
                clienteInstancia.nome, clienteInstancia.uriCliente, codigoProduto, nomeProduto, descricaoProduto, quantidadeProduto, precoUnidadeProduto, estoqueMinimoProduto)
            
            print(retorno)

        if opcao == '2':
            codigoProduto = input("\nQual o código do produto? ")
            quantidadeProduto = input("Qual a quantidade que deseja adicionar? ")

            retorno = servidorGerenciadorEstoque.adicionarProduto(codigoProduto, quantidadeProduto)

            print(retorno)
            

        if opcao == '3':
            codigoProduto = input("\nQual o código do produto? ")
            quantidadeProduto = input("Qual a quantidade do produto? ")
            print(servidorGerenciadorEstoque.retirarProduto(codigoProduto, quantidadeProduto))

        if opcao == '4':
            print(servidorGerenciadorEstoque.listarProdutos())

        if opcao == '5':
            print("\nAs opções de relatório são:")
            print("a - Relatório de produtos em estoque")
            print("b - Relatório de registros de movimentação de produtos")
            print("c - Relatório de produtos que acabaram")
            print("d - Relatório do fluxo de movimentação por período\n")
            #print("e - Relatório do produtos sem saída por período\n")
            opcaoRelatorio = input()

            if opcaoRelatorio == 'a':
                print(servidorGerenciadorEstoque.relatorioProdutosEstoque())

            if opcaoRelatorio == 'b':
                print(servidorGerenciadorEstoque.relatorioRegistros())

            if opcaoRelatorio == 'c':
                print(servidorGerenciadorEstoque.relatorioProdutosAcabaram())

            if opcaoRelatorio == 'd':
                dataInicio = input("\nQual a data de início do relatório? Digite no modelo: DD/MM/AAAA -  HH:MM:SS ")
                dataFim = input("Qual a data de fim do relatório? Digite no modelo: DD/MM/AAAA -  HH:MM:SS ")
                print(servidorGerenciadorEstoque.relatorioFluxoMovimentacao(dataInicio, dataFim))
"""
            if opcaoRelatorio == 'e':
                dataInicio = input("\nQual a data de início do relatório? Digite no modelo: DD/MM/AAAA -  HH:MM:SS ")
                dataFim = input("Qual a data de fim do relatório? Digite no modelo: DD/MM/AAAA -  HH:MM:SS ")
                servidorGerenciadorEstoque.relatorioProdutosSemSaida(clienteInstancia.nome, dataInicio, dataFim)
"""
                
            

File: cliente.js
const EventSource = require('eventsource');
const fetch = require('node-fetch');

const endpoint = 'http://127.0.0.1:5000/';

const eventSource = new EventSource(endpoint);
const readline = require('readline');


const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

//Tratamento de eventos SSE recebidos
// eventSource.addEventListener('message', event => {
//     const message = JSON.parse(event.data);
//     console.log('Nova mensagem recebida:', message);
// });


// eventSource.addEventListener('error', error => {
//     console.error('Erro na conexão SSE:', error);
// });

function cadastrarProduto() {
    rl.question("Digite o código do produto: ", (codigo) => {
        rl.question("Digite o nome do produto: ", (nome) => {
            rl.question("Digite a descrição do produto: ", (descricao) => {
                rl.question("Digite a quantidade do produto: ", (quantidade) => {
                    rl.question("Digite o preço por unidade do produto: ", (precoUnidade) => {
                        rl.question("Digite o estoque mínimo do produto: ", (estoqueMinimo) => {
                        const produto = {
                            codigo: codigo,
                            nome: nome,
                            descricao: descricao,
                            quantidade: quantidade,
                            precoUnidade: precoUnidade,
                            estoqueMinimo: estoqueMinimo,
                            acabou: 0,
                            estoqueBaixo: 0
                        };

                        fetch('http://127.0.0.1:5000/produto/novo', {
                            method: 'POST',
                            body: JSON.stringify(produto),
                            headers: { 'Content-Type': 'application/json' }
                        })
                            .then(response => response.json())
                            .then(data =>{
                                console.log(data);
                                iniciarCliente();
                            })
                            .catch(error => {
                                console.log("\nErro ao cadastrar produto:", error);
                                iniciarCliente();
                            
                            });
                        }
                        );
                    });
                });
            });
        });
    });
}

function adicionarProduto() {
    rl.question("Digite o código do produto: ", (codigo) => {
        rl.question("Digite a quantidade do produto: ", (qtdAdicionar) => {
            const produto = {
                codigo: codigo,
                qtdAdicionar: qtdAdicionar
            };

            fetch('http://127.0.0.1:5000/produto/adicionar', {
                method: 'POST',
                body: JSON.stringify(produto),
                headers: { 'Content-Type': 'application/json' }
            })
                .then(response => response.json())
                .then(data =>{
                    console.log(data);
                    iniciarCliente();
                })
                .catch(error => {
                    console.log("\nErro ao adicionar produto: ", error);
                    iniciarCliente();
                
                });  
        });
    });
}

function retirarProduto() {
    rl.question("Digite o código do produto: ", (codigo) => {
        rl.question("Digite a quantidade do produto: ", (qtdRetirar) => {
            const produto = {
                codigo: codigo,
                qtdRetirar: qtdRetirar
            };

            fetch('http://127.0.0.1:5000/produto/retirar', {
                method: 'POST',
                body: JSON.stringify(produto),
                headers: { 'Content-Type': 'application/json' }
            })
                .then(response => response.json())
                .then(data =>{
                    console.log(data);
                    iniciarCliente();
                }
                )
                .catch(error => {
                    console.log("\nErro ao retirar produto:", error);
                    iniciarCliente();

                });
        });
    });
}

function listarProdutos() {
    fetch('http://127.0.0.1:5000/produto/listar', {
        method: 'GET',
        headers: { 'Content-Type': 'application/json' }
    })
        .then(response => response.json())
        .then(data =>{
            console.log(data);
            iniciarCliente();
        }
        )
        .catch(error => {
            console.log("\nErro ao listar produtos:", error);
            iniciarCliente();

        });
}

function relatorioProdutosEmEstoque() {
    fetch('http://127.0.0.1:5000/produto/relatorio/estoque', {
        method: 'GET',
        headers: { 'Content-Type': 'application/json' }
    })
        .then(response => response.json())
        .then(data =>{
            console.log(data);
            iniciarCliente();
        }
        )
        .catch(error => {
            console.log("Erro ao listar produtos:", error);
            iniciarCliente();

        });
}

function relatorioMovimentacaoProdutos() {
    fetch('http://127.0.0.1:5000/produto/relatorio/movimento', {
        method: 'GET',
        headers: { 'Content-Type': 'application/json' }
    })
        .then(response => response.json())
        .then(data =>{
            console.log(data);
            iniciarCliente();
        }
        )
        .catch(error => {
            console.log("\nErro ao listar produtos:", error);
            iniciarCliente();

        });
}

function relatorioProdutosAcabaram() {
    fetch('http://127.0.0.1:5000/produto/relatorio/acabaram', {
        method: 'GET',
        headers: { 'Content-Type': 'application/json' }
    })
        .then(response => response.json())
        .then(data =>{
            console.log(data);
            iniciarCliente();
        }
        )
        .catch(error => {
            console.log("Erro ao listar produtos:", error);
            iniciarCliente();

        });
}

function relatorioFluxoMovimentacao() {
    rl.question("Qual a data de início do relatório? Digite no modelo: DD/MM/AAAA -  HH:MM:SS ", (dataInicio) => {
        rl.question("Qual a data de fim do relatório? Digite no modelo: DD/MM/AAAA -  HH:MM:SS ", (dataFim) => {
            const data = {
                dataInicio: dataInicio,
                dataFim: dataFim
            };
            
            fetch('http://127.0.0.1:5000/produto/relatorio/fluxo', {
                method: 'POST',
                body: JSON.stringify(data),
                headers: { 'Content-Type': 'application/json' }
            })
                .then(response => response.json())
                .then(data =>{
                    console.log(data);
                    iniciarCliente();
                }
                )
                .catch(error => {
                    console.log("Erro ao listar produtos:", error);
                    iniciarCliente();

                });
        });
    });
}


function exibirMenu() {
    console.log("\n===== Menu do Cliente =====");
    console.log("1 - Cadastrar novo produto");
    console.log("2 - Adicionar produto existente");
    console.log("3 - Retirar produto");
    console.log("4 - Listar produtos");
    console.log("5 - Gerar relatórios");
    console.log("0. Sair");
    console.log("===========================\n");
}

function exibirMenuRelatorio() {
    console.log("\n===== Menu de Relatórios =====");
    console.log("a - Relatório de produtos em estoque");
    console.log("b - Relatório de registros de movimentação de produtos");
    console.log("c - Relatório de produtos que acabaram");
    console.log("d - Relatório do fluxo de movimentação por período");
    console.log("x - Sair");
    console.log("===========================\n");
}

function processarEscolhaRelatorio(escolha) {
    switch (escolha) {
        case 'a':
            relatorioProdutosEmEstoque();
            break;
        case 'b':
            relatorioMovimentacaoProdutos();
            break;
        case 'c':
           relatorioProdutosAcabaram();
            break;
        case 'd':
            relatorioFluxoMovimentacao();
            break;
        case  'x':
            iniciarCliente();
            break;
        default:
            console.log("\nEscolha inválida. Tente novamente.\n");
            iniciarCliente();
            break;
    }
}

function processarEscolha(escolha) {
    switch (escolha) {
        case '1':
            cadastrarProduto();
            break;
        case '2':
            adicionarProduto();
            break;
        case '3':
            retirarProduto();
            break;
        case '4':
            listarProdutos();
            break;

        case '5':
            exibirMenuRelatorio();
            rl.question("Digite sua escolha: ", (escolha) => {
                // Obtém o primeiro caractere da escolha
                const primeiroCaractere = escolha.charAt(0);

                // Chame a função processarEscolha com o primeiro caractere
                processarEscolhaRelatorio(primeiroCaractere);

                // if (primeiroCaractere !== '0') {
                //     iniciarCliente();
                // }
            });
            break;

        case '0':
            rl.close();
            break;
            
        default:
            console.log("\nEscolha inválida. Tente novamente.\n");
            iniciarCliente();
            break;
    }
}


function iniciarCliente() {
    exibirMenu();

    rl.question("Digite sua escolha: ", (escolha) => {
        // Obtém o primeiro caractere da escolha
        const primeiroCaractere = escolha.charAt(0);

        // Chame a função processarEscolha com o primeiro caractere
        processarEscolha(primeiroCaractere);

        // console.log("Teste");

        // if (primeiroCaractere !== '0') {
        //     iniciarCliente();
        // }
    });
}

iniciarCliente();